ComponentKit/Utilities/CKComponentAction.mm

```diff
@@ -50,35 +50,45 @@ void CKConfigureInvocationWithArguments(NSInvocation *invocation, NSInteger inde
     case CKTypedComponentActionVariant::TargetSelector:
       return _target;
     case CKTypedComponentActionVariant::Responder:
-      return _scopedResponder.responder;
+      return _scopeIDAndGenerator.second();
     case CKTypedComponentActionVariant::Block:
       CKCFailAssert(@"Should not be asking for target for block action.");
       return nil;
   }
 }
 
-CKTypedComponentActionBase::CKTypedComponentActionBase() noexcept : _target(nil), _scopedResponder(nil), _block(NULL), _variant(CKTypedComponentActionVariant::RawSelector), _selector(nullptr) {}
+CKTypedComponentActionBase::CKTypedComponentActionBase() noexcept : _target(nil), _scopeIDAndGenerator({}), _block(NULL), _variant(CKTypedComponentActionVariant::RawSelector), _selector(nullptr) {}
 
-CKTypedComponentActionBase::CKTypedComponentActionBase(id target, SEL selector) noexcept : _target(target), _scopedResponder(nil), _block(NULL), _variant(CKTypedComponentActionVariant::TargetSelector), _selector(selector) {};
+CKTypedComponentActionBase::CKTypedComponentActionBase(id target, SEL selector) noexcept : _target(target), _scopeIDAndGenerator({}), _block(NULL), _variant(CKTypedComponentActionVariant::TargetSelector), _selector(selector) {};
 
-CKTypedComponentActionBase::CKTypedComponentActionBase(const CKComponentScope &scope, SEL selector) noexcept : _target(nil), _scopedResponder([scope.scopeHandle() scopedResponder]),  _block(NULL), _variant(CKTypedComponentActionVariant::Responder), _selector(selector)
+CKTypedComponentActionBase::CKTypedComponentActionBase(const CKComponentScope &scope, SEL selector) noexcept : _target(nil), _block(NULL), _variant(CKTypedComponentActionVariant::Responder), _selector(selector)
 {
-  CKCAssert(_scopedResponder, @"You are creating an action that will not fire because you have an invalid scope handle.");
+  const auto handle = scope.scopeHandle();
+  CKCAssert(handle, @"You are creating an action that will not fire because you have an invalid scope handle.");
+
+  const auto scopedResponder = handle.scopedResponder;
+  const auto responderKey = [scopedResponder keyForHandle:handle];
+  _scopeIDAndGenerator = {
+    [handle globalIdentifier],
+    ^id(void) {
+      return [scopedResponder responderForKey:responderKey];
```

Comment:
A few comments here would be helpful as well. Particularly mentioning the need for the responder key and why this generator block is necessary.